Project:  Identify malarial parasite in blood cells in resource poor settings from microscopic pictures taken by cellphone. 

         Create model on a regular laptop without GPU
         
         Deploy the model in a phone on on the web
         
         
inspiration: https://blog.insightdatascience.com/https-blog-insightdatascience-com-malaria-hero-a47d3d5fc4bb by Carlos Atico Ariza


In [1]:

# importing turicreate as tc

import warnings
# ignoring warnings
warnings.filterwarnings("ignore", message="numpy.dtype size changed")
warnings.filterwarnings("ignore", message="numpy.ufunc size changed")

import turicreate as tc

Data source: Data was obtained from https://ceb.nlm.nih.gov/repositories/malaria-datasets/

There is a link to download the files at the bottom of the website. 

Download and unzip the files to your working directory for this Jupyter notebook

In [4]:
# Loading the images from local drive.

reference_data  = tc.image_analysis.load_images('cell_images', with_path=True)

Read 9996 images in 5.16255 secs	speed: 1936.25 file/sec

Read 21178 images in 10.1628 secs	speed: 2236.29 file/sec

In [5]:
# checking the data 
reference_data.head()

Columns:
	path	str
	image	Image

Rows: 10

Data:
+-------------------------------+------------------------+
|              path             |         image          |
+-------------------------------+------------------------+
| cell_images/Parasitized/C1... | Height: 148 Width: 142 |
| cell_images/Parasitized/C1... | Height: 208 Width: 148 |
| cell_images/Parasitized/C1... | Height: 151 Width: 139 |
| cell_images/Parasitized/C1... | Height: 154 Width: 145 |
| cell_images/Parasitized/C1... | Height: 154 Width: 133 |
| cell_images/Parasitized/C1... | Height: 145 Width: 169 |
| cell_images/Parasitized/C1... | Height: 139 Width: 139 |
| cell_images/Parasitized/C1... | Height: 124 Width: 154 |
| cell_images/Parasitized/C1... | Height: 148 Width: 124 |
| cell_images/Parasitized/C1... | Height: 142 Width: 139 |
+-------------------------------+------------------------+
[10 rows x 2 columns]

In [6]:
image_sarray = reference_data['image']

In [7]:
# Resizing the image -you can skip this 

resized_images = tc.image_analysis.resize(image_sarray,224,224)

In [8]:
# Reinstating the edited "image" column back to reference_data

reference_data['image'] = resized_images

In [9]:
# Labelling each images in the SFrame using the folder name as malaria or not malaria 

reference_data['label'] = reference_data['path'].apply(lambda path: 'malaria' if '/Parasitized' in path else 'not malaria')


In [10]:
# Displaying first 5 rows to make sure that we got the data in the right format

reference_data.head()

Columns:
	path	str
	image	Image
	label	str

Rows: 10

Data:
+-------------------------------+------------------------+---------+
|              path             |         image          |  label  |
+-------------------------------+------------------------+---------+
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
| cell_images/Parasitized/C1... | Height: 224 Width: 224 | malaria |
+-------------------------------+------------------------+---------+
[10 rows x 3 columns]

In [11]:
#Splitting the data into training, testing and validation sets

train_data2, test_data = reference_data.random_split(0.8)
train_data, valid_data = train_data2.random_split(0.8)

In [12]:
# Create model
model = tc.image_classifier.create(train_data, target= 'label', max_iterations = 30,validation_set = valid_data  )

Performing feature extraction on resized images...
Completed    64/17616
Completed   128/17616
Completed   192/17616
Completed   256/17616
Completed   320/17616
Completed   384/17616
Completed   448/17616
Completed   512/17616
Completed   576/17616
Completed   640/17616
Completed   704/17616
Completed   768/17616
Completed   832/17616
Completed   896/17616
Completed   960/17616
Completed  1024/17616
Completed  1088/17616
Completed  1152/17616
Completed  1216/17616
Completed  1280/17616
Completed  1344/17616
Completed  1408/17616
Completed  1472/17616
Completed  1536/17616
Completed  1600/17616
Completed  1664/17616
Completed  1728/17616
Completed  1792/17616
Completed  1856/17616
Completed  1920/17616
Completed  1984/17616
Completed  2048/17616
Completed  2112/17616
Completed  2176/17616
Completed  2240/17616
Completed  2304/17616
Completed  2368/17616
Completed  2432/17616
Completed  2496/17616
Completed  2560/17616
Completed  2624/17616
Completed  2688/17616
Completed  2752/17616
Com

Logistic regression:

--------------------------------------------------------

Number of examples          : 17616

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 2048

Number of coefficients      : 2049

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 6        | 0.000003  | 3.046645     | 0.502498          | 0.498546            |

| 2         | 8        | 1.000000  | 4.436010     | 0.822377          | 0.821293            |

| 3         | 9        | 1.000000  | 5.526598     | 0.844630          | 0.849027            |

| 4         | 10       | 1.000000  | 6.607287     | 0.842757          | 0.846119            |

| 5         | 11       | 1.000000  | 7.628104     | 0.857232          | 0.857750            |

| 10        | 16       | 1.000000  | 12.754394    | 0.876306          | 0.876985            |

| 15        | 21       | 1.000000  | 17.848421    | 0.919051          | 0.919481            |

| 20        | 27       | 1.000000  | 23.323171    | 0.928588          | 0.929546            |

| 25        | 33       | 1.000000  | 28.879053    | 0.931994          | 0.931335            |

| 30        | 39       | 1.000000  | 34.358547    | 0.939089          | 0.931783            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

Completed (Iteration limit reached).

This model may not be optimal. To improve it, consider increasing `max_iterations`.

In [13]:
# storing the prediction 

predictions = model.predict(test_data)

In [14]:
# Evaluating the model on the test data

metrics = model.evaluate(test_data)
print(metrics['accuracy'])

Performing feature extraction on resized images...
Completed   64/5471
Completed  128/5471
Completed  192/5471
Completed  256/5471
Completed  320/5471
Completed  384/5471
Completed  448/5471
Completed  512/5471
Completed  576/5471
Completed  640/5471
Completed  704/5471
Completed  768/5471
Completed  832/5471
Completed  896/5471
Completed  960/5471
Completed 1024/5471
Completed 1088/5471
Completed 1152/5471
Completed 1216/5471
Completed 1280/5471
Completed 1344/5471
Completed 1408/5471
Completed 1472/5471
Completed 1536/5471
Completed 1600/5471
Completed 1664/5471
Completed 1728/5471
Completed 1792/5471
Completed 1856/5471
Completed 1920/5471
Completed 1984/5471
Completed 2048/5471
Completed 2112/5471
Completed 2176/5471
Completed 2240/5471
Completed 2304/5471
Completed 2368/5471
Completed 2432/5471
Completed 2496/5471
Completed 2560/5471
Completed 2624/5471
Completed 2688/5471
Completed 2752/5471
Completed 2816/5471
Completed 2880/5471
Completed 2944/5471
Completed 3008/5471
Completed

In [24]:
# Printing confusion matrix

print ("Confusion Matrix : \n%s" % metrics['confusion_matrix'])

Confusion Matrix : 
+--------------+-----------------+-------+
| target_label | predicted_label | count |
+--------------+-----------------+-------+
|   malaria    |     malaria     |  2538 |
|   malaria    |   not malaria   |  203  |
| not malaria  |   not malaria   |  2564 |
| not malaria  |     malaria     |  166  |
+--------------+-----------------+-------+
[4 rows x 3 columns]



In [16]:
model.summary()

Class                                    : ImageClassifier

Schema
------
Number of classes                        : 2
Number of feature columns                : 1
Input image shape                        : (3, 224, 224)

Training summary
----------------
Number of examples                       : 17616
Training loss                            : 2974.4021
Training time (sec)                      : 3490.1657



In [ ]:
#It took 58 minutes to complete model generation from 22,147 images on a laptop with 
# Intel core i7-6500U 2.5GHZ , 16 gb memory without using the GPU

In [17]:
 #Save the model for later use in Turi Create

model.save('malaria.model')

In [18]:
#Export for use in Core ML
# You can use this model in the template provided by apple to create an iphone app
# Refer to this website if you need assistance - https://www.appcoda.com/core-ml-model-with-python/

model.export_coreml('malaria.mlmodel')

In [19]:
print(metrics['f1_score'])

0.932872475896


In [20]:
targets = test_data['label']

In [21]:
f1 = tc.evaluation.f1_score(targets, predictions)

In [22]:
fbeta = tc.evaluation.fbeta_score(targets, predictions, beta = 2.0)

In [23]:
print(f1, fbeta)

(0.9328724758959431, 0.9366552202820194)


Stats

F-beta score = 93.6

Accuracy = 93.26%

Sensitivity = 92.5%

Specificity = 93.2%

Positive predictive value = 93.8%

Negative predictive value = 92.6%
